lib/Onyx.d.ts

```diff
@@ -0,0 +1,278 @@
+import {Component} from 'react';
+import {PartialDeep} from 'type-fest';
+import * as Logger from './Logger';
+import {DeepRecord, Key, CollectionKey, Selector, Value} from './types';
+
+type KeyValueMap = {
+    [TKey in Key | CollectionKey]: Value[TKey];
+};
+
+type BaseConnectOptions<TKey extends Key | CollectionKey> = {
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+    selector?: Selector<TKey, Value[TKey] | null>;
+};
+
+type ConnectOptions<TKey extends Key | CollectionKey> = BaseConnectOptions<TKey> &
+    (
+        | {
+              key: TKey extends CollectionKey ? TKey : never;
+              callback?: (value: Record<string, Value[TKey]> | null, key?: TKey) => void;
+              waitForCollectionCallback: true;
+          }
+        | {
+              key: TKey;
+              callback?: (value: Value[TKey] | null, key?: TKey) => void;
+              waitForCollectionCallback?: false;
+          }
+    );
+
+type MergeCollection<TKey extends Key, TMap, TValue> = {
+    [MapK in keyof TMap]: MapK extends `${TKey}${string}`
+        ? MapK extends `${TKey}`
+            ? never // forbids empty id
+            : TValue
+        : never;
+};
+
+type UpdateOperation<TKey extends Key | CollectionKey> =
+    | {
+          onyxMethod: 'set';
+          key: TKey;
+          value: Value[TKey];
+      }
+    | {
+          onyxMethod: 'merge';
+          key: TKey;
+          value: PartialDeep<Value[TKey]>;
+      }
+    | {
+          [TKey in CollectionKey]: {
+              onyxMethod: 'mergeCollection';
+              key: TKey;
+              value: Record<`${TKey}${string}`, PartialDeep<Value[TKey]>>;
+          };
+      }[CollectionKey];
+
+type UpdateOperations<TKeyList extends (Key | CollectionKey)[]> = {
+    [TKey in keyof TKeyList]: UpdateOperation<TKeyList[TKey]>;
+};
+
+type InitOptions = {
+    keys?: DeepRecord<string, string>;
+    initialKeyStates?: Partial<KeyValueMap>;
+    safeEvictionKeys?: Key[];
+    maxCachedKeysCount?: number;
+    captureMetrics?: boolean;
+    shouldSyncMultipleInstances?: boolean;
+    debugSetState?: boolean;
+};
+
+declare const METHOD: {
+    readonly SET: string;
+    readonly MERGE: string;
+    readonly MERGE_COLLECTION: string;
+    readonly CLEAR: string;
```

Comment:
should these be literal strings?

Reply:
I think so, let's change it to literal strings @fabioh8010!